apis/apps/v1alpha1/workloadspread_types.go

```diff
@@ -0,0 +1,204 @@
+	// redundant capacity after the subset where the Pod lives. If a Pod was scheduled failed and still in a unschedulabe status
+	// over RescheduleCriticalSeconds duration, the controller will reschedule it to a suitable subset.
+	// +optional
+	RescheduleCriticalSeconds *int32 `json:"rescheduleCriticalSeconds,omitempty"`
+}
+
+// WorkloadSpreadSubset defines the details of a subset.
+type WorkloadSpreadSubset struct {
+	// Name should be unique between all of the subsets under one WorkloadSpread.
+	Name string `json:"name"`
+
+	// Indicates the node required selector to form the subset.
+	// +optional
+	RequiredNodeSelectorTerm *corev1.NodeSelectorTerm `json:"requiredNodeSelectorTerm,omitempty"`
+
+	// Indicates the node preferred selector to form the subset.
+	// +optional
+	PreferredNodeSelectorTerms []corev1.PreferredSchedulingTerm `json:"preferredNodeSelectorTerms,omitempty"`
+
+	// Indicates the tolerations the pods under this subset have.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// MaxReplicas indicates the desired max replicas of this subset.
+	// +optional
+	MaxReplicas *intstr.IntOrString `json:"maxReplicas,omitempty"`
+
+	// Patch indicates the patch metadata to the Pod.
+	// +optional
+	Patch runtime.RawExtension `json:"patch,omitempty"`
+}
+
+// WorkloadSpreadStatus defines the observed state of WorkloadSpread.
+type WorkloadSpreadStatus struct {
+	// ObservedGeneration is the most recent generation observed for this WorkloadSpread. It corresponds to the
+	// WorkloadSpread's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Contains the status of each subset. Each element in this array represents one subset
+	// +optional
+	SubsetStatuses []WorkloadSpreadSubsetStatus `json:"subsetStatuses,omitempty"`
+}
+
+// SubsetUnscheduledStatus contains the details for the unscheduled subset.
+type SubsetUnscheduledStatus struct {
+	// LastScheduleFailureTime indicates the last time for unscheduled subset.
+	// It's valid only when the WorkloadSpreadScheduleStrategy is 'adaptive', otherwise it is always null.
+	// When one or more one pods in a subset have condition[PodScheduled] = false, the subset is considered temporarily unschedulable.
+	// The condition[PodScheduled] = false, which means the cluster does not have enough resources to schedule at this moment,
+	// even if only single Pod scheduled fails. Controller will set this field with a unschedulable time.
+	// The LastScheduleFailureTime is valid in a period of time(e.g. 5m), and the webhook will skip this subset by checking whether time.Now() < LastScheduleFailureTime + 5m.
+	// After a period of time，the controller will set this filed to empty, indicating that it is schedulable, and the webhook can't ignore this subset.
+	// +optional
+	LastScheduleFailureTime *metav1.Time `json:"lastScheduleFailureTime,omitempty"`
```

Comment:
plz consider make the status fields more extensible(replace fixed schedule condition with conditions array), so that we can record other status in the subset here, for example, we can record that creating a pods in a subset will be denied by the quota

Reply:
+1 to add condition array